### PR TITLE
ApplySchema: allow-zero-in-date embedded as query comment in call to ExecuteFetchAsDba

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Fixes https://github.com/vitessio/vitess/issues/9996

Previously, `vtctlclient ApplySchema -skip_preflight -strategy "direct -allow-zero-in-date"` would prepend a given SQL with a `set @@session.sql_mode=` before sending to `ExecuteFetchAsDba`. This caused errors in the SQL to be silently dropped.

Now, instead, we embed a `/*vt+ allowZeroInDate=true */` hint in the SQL (only applicable to DDLs, which is fine, this is all we need).

Then, on `tabletmaneger`'s side, we parse the query, and if we're able to find such `/*vt+ allowZeroInDate=true */`, we run `set @@session.sql_mode=` before actually executing the SQL.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/9996
- caused by #9273 

Added validation test

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletmanager/rpc_query.go [hunk 1]
```diff
@@ -49,6 +50,23 @@ func (tm *TabletManager) ExecuteFetchAsDba(ctx context.Context, query []byte, db
...
+	if stmt, err := sqlparser.Parse(string(query)); err == nil {
+		if ddlStmt, ok := stmt.(sqlparser.DDLStatement); ok {
+			if comments := ddlStmt.GetComments(); comments != nil {
+				directives := sqlparser.ExtractCommentDirectives(comments)
+				if _, ok := directives["allowZeroInDate"]; ok {
+					allowZeroInDate = true
+				}
+			}
+		}
+	}
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletmanager/rpc_query.go:64]: nit: kindly reduce the cyclomatic complexity

File: go/vt/vttablet/tabletmanager/rpc_query.go [hunk 2]
```diff
@@ -49,6 +50,23 @@ func (tm *TabletManager) ExecuteFetchAsDba(ctx context.Context, query []byte, db
...
+	if allowZeroInDate {
+		if _, err := conn.ExecuteFetch("set @@session.sql_mode=REPLACE(REPLACE(@@session.sql_mode, 'NO_ZERO_DATE', ''), 'NO_ZERO_IN_DATE', '')", 1, false); err != nil {
+			return nil, err
+		}
+	}
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletmanager/rpc_query.go:69]: I think this can be done inline where the boolean is set.

**Review by harshit-gangal — APPROVED** [review]
overall looks good. some nits.

File: go/vt/vttablet/tabletmanager/rpc_query.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletmanager/rpc_query.go]: @harshit-gangal I've turned it into something else; I don't think the rewrite is any better; the original made more sense to me. Please let me know what you think.

File: go/vt/vttablet/tabletmanager/rpc_query.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletmanager/rpc_query.go]: done

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vttablet/tabletmanager/rpc_query.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vttablet/tabletmanager/rpc_query.go]: I felt it is easier to follow.

**Pull request merged**